lib/pushbox.js

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const isA = require('joi')
+const Pool = require('./pool')
+const P = require('./promise')
+const {URL} = require('url')
+const validators = require('./routes/validators')
+
+const LOG_OP_RETRIEVE = 'pushbox.retrieve'
+const LOG_OP_STORE = 'pushbox.store'
+
+const SERVICE_FXA = 'fxa'
+const TOPICS_TTL = {
+  sendtab: 30 * 24 * 3600 // 30 days.
+}
+
+const PUSHBOX_RETRIEVE_SCHEMA = isA.alternatives().try(
+  // common case
+  isA.object({
+    last: isA.boolean().required(),
+    index: isA.number().required(),
+    messages: isA.array().items(isA.object({
+      index: isA.number().required(),
+      data: isA.string().regex(validators.URL_SAFE_BASE_64).required(),
+    })).optional(),
+  }),
+  // limit == 0 only returns the last index.
+  isA.object({
+    index: isA.number().required(),
+  }),
+  // error case
+  isA.object({
+    error: isA.string().required(),
+    status: isA.number().required(),
+  })
+)
+
+module.exports = function (log, config) {
+  const pool = new Pool(config.pushboxUrl, { timeout: 15000 })
```

Comment:
While we roll this out, I think it would be worth doing a little feature-flag here.  You could add a `config.pushboxEnabled` boolean flag, or a special-case value like `config.pushboxUrl === "none"`.  Either way, if you detect that pushbox is not enabled, you can return a little stub implementation like:

```
if (! config.pushboxEnabled) {
  return {
    retreive: () => { throw errors.featureNotEnabled() },
    store: () => { throw errors.featureNotEnabled() }
}
```

This will allow us to merge before production environment is ready, and to easily pref it off if we run into problems in production.  (Of course, clients would also need to know that a "feature not enabled" error is a possibility).

If you wanted to get *really* fancy, you could add the ability to gradually roll it out to a proportion of users based on uid, using the sampling facilities offered in features.js here:

  https://github.com/mozilla/fxa-auth-server/blob/master/lib/features.js

But a simple on/off switch should suffice for initial landing.